utils/altcoinService.ts

```diff
@@ -115,3 +115,45 @@ export const getAutoRenewAllowance = (
 
   return allowance;
 };
+
+type AvnuQuote = {
+  address: string;
+  currentPrice: number;
+  decimals: number;
+};
+
+export const smartCurrencyChoosing = async (
```

Comment:
I think this function is a bit difficult to understand as one block, I asked chatGPT 4 to refactor it, here is a proposition he made: 

```ts
enum CurrencyType {
  ETH = 'ETH',
  STRK = 'STRK'
}

interface TokenBalances {
  ETH: string;
  STRK: string;
}

interface AvnuQuote {
  address: string;
  currentPrice: number;
  decimals: number;
}

// Helper to convert hexadecimal to decimal
const hexToDecimal = (hex: string): number => parseInt(hex, 16);

// API call to fetch quote data
async function fetchAvnuQuoteData(): Promise<AvnuQuote[]> {
  const response = await fetch(`${process.env.NEXT_PUBLIC_AVNU_API}/tokens/short?in=0x49d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7`);
  return response.json();
}

// Determine the currency type based on token balances and STRK quote
async function determineCurrencyType(tokenBalances: TokenBalances): Promise<CurrencyType> {
  // Early returns based on token presence
  if (tokenBalances.ETH && !tokenBalances.STRK) return CurrencyType.ETH;
  if (!tokenBalances.ETH) return CurrencyType.STRK;

  // Fetch data and find relevant quote
  const avnuQuotes = await fetchAvnuQuoteData();
  const contractAddress = ERC20Contract[CurrencyType.STRK];
  const quoteData = avnuQuotes.find(quote => hexToDecimal(quote.address) === hexToDecimal(contractAddress));

  // Fallback to ETH if no STRK quote data is found
  if (!quoteData) return CurrencyType.ETH;

  // Convert STRK balance to ETH equivalent
  const strkBalance = BigInt(tokenBalances.STRK);
  const strkQuote = BigInt(Math.round(quoteData.currentPrice * Math.pow(10, quoteData.decimals)));
  const strkConvertedBalance = (strkBalance * strkQuote) / BigInt(Math.pow(10, quoteData.decimals));

  // Compare converted STRK balance to ETH balance
  return BigInt(tokenBalances.ETH) * BigInt(3) < strkConvertedBalance ? CurrencyType.STRK : CurrencyType.ETH;
}

// Example usage:
const tokenBalances: TokenBalances = { ETH: '100', STRK: '200' };
determineCurrencyType(tokenBalances).then(currency => console.log(currency));

```

I think it's way cleaner :), and it'll permit you to have easier tests with better coverage! Don't forget to test :)